Fix cargo install --index when used with registry.default

arlosi opened the pull request:
Setting `registry.default` causes the `args.registry` call to return the default registry as if it were passed through `--registry`, which leaves the `--index` argument ignored in `cargo install`, since `registry` is checked first.

Fixes #11301 by checking for `index` before `registry`.

Note that if you try to pass both `--index` and `--registry`, then a command-line parser error (correctly) occurs:
```
The argument '--registry <REGISTRY>' cannot be used with '--index <INDEX>'
```

Eh2406 commented:
You have pointed out before that `registry.default` is woefully under tested. Would this be a good PR to try to fix that?

weihanglo reviewed (approved):
This looks fine. And I agree with Eh2406. We might want some tests against `registry.default`. Or at least posting an issue calling out for help.

weihanglo commented:
A bit off-topic. I wonder if `--index` and `--registry` are mutually exclusive[^1] throughout all commands in Cargo. If that is true, maybe we should have an enum combining them during arg parsing instead passing down two separate `Option<&str>`.

[^1]: At least for some registry operations, it seems to be the case. https://github.com/rust-lang/cargo/blob/67398aec95a7eccb0cfa2347fc60dcbcdd597b6c/src/cargo/ops/registry.rs#L1032-L1039

weihanglo commented:
Friendly ping @arlosi. Do you want to add a test for this particular scenario? A test following steps described in #11301 seems fine.

arlosi commented:
Added a test and changed the implementation strategy slightly.

There are complications when trying to add a new `RegistryOrIndex` enum type. `--registry` and `--index` do seem to be mutually exclusive, but they're not always both used. For example `cargo new` only uses `--registry` and `cargo publish` does additional processing of the `registry` value based on the `publish` field in `Cargo.toml` (if set).

weihanglo reviewed (approved):
I am not entirely sold but yup we can proceed first. I'll create an issue for exploring a way to make it more future-proof. Thank you for fixing this :)

weihanglo commented:
@bors r+

bors commented:
:pushpin: Commit 24bf873c83a902c1a26c0b7e1ab0df64c96135ef has been approved by `weihanglo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=weihanglo 24bf873c83a902c1a26c0b7e1ab0df64c96135ef -->
<!-- homu: {"type":"Approved","sha":"24bf873c83a902c1a26c0b7e1ab0df64c96135ef","approver":"weihanglo","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 24bf873c83a902c1a26c0b7e1ab0df64c96135ef with merge b690ab48fc079bf068fcf0352096ddfe5821269f...
<!-- homu: {"type":"BuildStarted","head_sha":"24bf873c83a902c1a26c0b7e1ab0df64c96135ef","merge_sha":"b690ab48fc079bf068fcf0352096ddfe5821269f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3479154693/jobs/5818135498)
Approved by: weihanglo
Pushing b690ab48fc079bf068fcf0352096ddfe5821269f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"weihanglo","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/3479154693/jobs/5818135498"},"merge_sha":"b690ab48fc079bf068fcf0352096ddfe5821269f"} -->
